app/src/main/kotlin/compiler/parser/Parser.kt

```diff
@@ -148,26 +148,22 @@ class Parser<S : Comparable<S>>(
                     }
 
                     is ParserAction.Reduce -> {
-                        val start: Location
-                        val end: Location
 
-                        if (parsedSubtrees.isNotEmpty()) {
-                            start = parsedSubtrees.first().location.start
-                            end = parsedSubtrees.last().location.end
+                        val location: LocationRange = if (parsedSubtrees.isNotEmpty()) {
+                            LocationRange(parsedSubtrees.first().location.start, parsedSubtrees.last().location.end)
                         } else {
                             // An empty non-terminal symbol has (somewhat arbitrarily) the location of the next input symbol.
-                            start = lookaheadStart
-                            end = lookaheadEnd
+                            LocationRange(lookaheadStart, lookaheadEnd)
                         }
 
-                        callResult = ParseTree.Branch(LocationRange(start, end), symbol, parsedSubtrees, action.production)
+                        callResult = ParseTree.Branch(location, symbol, parsedSubtrees, action.production)
                         callStack.removeLast()
                     }
 
                     null -> {
                         // Report a parsing error when the parsing table does not tell what to do.
-                        val expectedSymbols = parseActions.keys.filter { it.first == dfa && it.second == state }.mapNotNull { it.third }
-                        diagnostics.report(Diagnostic.ParserError(lookahead?.symbol, lookaheadStart, lookaheadEnd, expectedSymbols))
+                        // val expectedSymbols = parseActions.keys.filter { it.first == dfa && it.second == state }.mapNotNull { it.third }
+                        diagnostics.report(Diagnostic.ParserError.UnexpectedToken(lookahead?.symbol, LocationRange(lookaheadStart, lookaheadEnd), /*expectedSymbols*/))
```

Comment:
Why no longer telling the expected symbols?

Reply:
Usually, the list of expected symbols is very long and not user-friendly. Some EXPR512 and others. This has no meaning if you don't know the grammar details. Maybe, we can add a task to tidy this up.

Reply:
Perhapse a good idea would be to display the *tokens* that these non-terminal symbols could begin with (the FIRST set from the grammar analysis).

Reply:
Is this conversation resolved?

Reply:
I assume that we stay with what we have right now, we'll surely discuss that matter with whole team.

Reply:
I think, that Tomek's idea is very nice, also sounds simple, so maybe I can implement it within this PR.